### PR TITLE
better extra attributes

### PR DESCRIPTION
**Pull request by maxiloc**

### CODE REVIEW

File: scraper/src/strategies/default_strategy.py [hunk 1]
```diff
@@ -190,24 +195,36 @@ def _generate_empty_hierarchy():
...
         # Handle default values
         for level in selectors:
-            if level != 'content':
+            if level in levels:
                 if hierarchy[level] is None and selectors[level]['default_value'] is not None:
                     hierarchy[level] = selectors[level]['default_value']
-            else:
+            elif level == 'content':
                 if content is None and selectors[level]['default_value'] is not None:
                     content = selectors['content']['default_value']
```
Comment (s-pace) [inline scraper/src/strategies/default_strategy.py]: else? should we warn it?

File: scraper/src/strategies/default_strategy.py [hunk 2]
```diff
@@ -190,24 +195,36 @@ def _generate_empty_hierarchy():
...
+    def _handle_default_values(hierarchy, content, selectors, levels):
         # Handle default values
         for level in selectors:
-            if level != 'content':
+            if level in levels:
                 if hierarchy[level] is None and selectors[level]['default_value'] is not None:
                     hierarchy[level] = selectors[level]['default_value']
-            else:
+            elif level == 'content':
                 if content is None and selectors[level]['default_value'] is not None:
```
Comment (s-pace) [inline scraper/src/strategies/default_strategy.py]: Regarding content: None or empty too no?

File: scraper/src/strategies/default_strategy.py [hunk 3]
```diff
@@ -190,24 +195,36 @@ def _generate_empty_hierarchy():
...
                     content = selectors['content']['default_value']
 
         return hierarchy, content
 
+    def _get_nodes_per_global_level(self, selectors, levels):
+        for level in selectors.keys():
+            level_selector = selectors[level]
+            if level not in levels or level_selector['global']:
```
Comment (s-pace) [inline scraper/src/strategies/default_strategy.py]: better to check a boolean first? iterating cost more (short-circuit)

File: scraper/src/strategies/default_strategy.py [hunk 4]
```diff
@@ -190,24 +195,36 @@ def _generate_empty_hierarchy():
...
+                matching_dom_nodes = self.select(level_selector['selector'])
+                self.global_content[level] = self.get_text_from_nodes(matching_dom_nodes,
+                                                                      self.get_strip_chars(level, selectors))
+
+                if self.global_content[level] is None and level_selector['default_value'] is not None:
```
Comment (s-pace) [inline scraper/src/strategies/default_strategy.py]: same, should we check size ?

File: scraper/src/strategies/default_strategy.py [hunk 5]
```diff
@@ -233,7 +246,7 @@ def _get_selector_all(levels, selectors):
         for level in levels:
             level_selector = selectors[level]
 
-            if len(level_selector['selector']) > 0:
+            if len(level_selector['selector']) > 0 and not level_selector['global']:
```
Comment (s-pace) [inline scraper/src/strategies/default_strategy.py]: short-circuit, better the other way around (nested method)

File: scraper/src/strategies/abstract_strategy.py [hunk 6]
```diff
@@ -134,6 +134,9 @@ def get_text(element, strip_chars=None):
     def get_text_from_nodes(elements, strip_chars=None):
         """Return the text content of a set of DOM nodes"""
 
+        if not isinstance(elements, list):
+            return elements
```
[the remark this answers is not in the document]
Comment (maxiloc) [reply scraper/src/strategies/abstract_strategy.py]: 👍

File: scraper/src/strategies/default_strategy.py (hunk 2 above)
Comment (maxiloc) [reply scraper/src/strategies/default_strategy.py]: 👍

File: scraper/src/strategies/default_strategy.py (hunk 1 above)
Comment (maxiloc) [reply scraper/src/strategies/default_strategy.py]: 👍

File: scraper/src/strategies/default_strategy.py (hunk 3 above)
Comment (maxiloc) [reply scraper/src/strategies/default_strategy.py]: 👍

File: scraper/src/strategies/default_strategy.py (hunk 4 above)
Comment (maxiloc) [reply scraper/src/strategies/default_strategy.py]: 👍

File: scraper/src/strategies/default_strategy.py (hunk 5 above)
Comment (maxiloc) [reply scraper/src/strategies/default_strategy.py]: 👍

File: scraper/src/strategies/default_strategy.py [hunk 7]
```diff
@@ -104,6 +104,9 @@ def get_records_from_dom(self, current_page_url=None):
             # We only save content for the 'text' matches
             content = None if current_level != 'content' else self.get_text(node, self.get_strip_chars(current_level, selectors))
 
+            if (content is None or content == "") and current_level == 'content':
```
Comment (s-pace) [inline scraper/src/strategies/default_strategy.py]: You could even use bool(content)
https://infohost.nmt.edu/tcc/help/pubs/python/web/bool-function.html

**Pull request merged**